#**ВЕБИНАР 07.05.2024 Занятие по AutoKeras, созданное с помощью ChatGPT**

# **1. Обзор занятия**

Добро пожаловать на занятие по AutoKeras для начинающих! Наша цель на сегодняшнем занятии - предоставить вам полное понимание основ AutoKeras, мощного инструмента для автоматического машинного обучения, который может значительно упростить и ускорить процесс создания моделей машинного обучения.


## Цели занятия:

- Познакомить вас с основными концепциями и возможностями AutoKeras.
- Обсудить различные модули AutoKeras, включая обработку изображений, табличных данных, текстов и временных рядов.
- Провести серию практических заданий, которые помогут закрепить теоретические знания на практике.

## Структура занятия:

1. Введение в AutoKeras и обзор его возможностей.
2. Теоретическая часть: подробное изучение работы с различными типами данных.
3. Практические задания: прогнозирование объема продаж, классификация отзывов на Tesla и классификация рентгеновских снимков.
4. Итоги занятия и ответы на вопросы.

Надеемся, что это занятие будет информативным и познавательным, и вы сможете применить полученные знания в своих проектах!

# **2. Основы AutoKeras**

AutoKeras — это мощная библиотека для автоматизированного машинного обучения (AutoML), разработанная для того, чтобы сделать машинное обучение более доступным и эффективным. Этот инструмент позволяет автоматически находить оптимальные модели для различных задач машинного обучения, сокращая время и усилия, необходимые для ручного тюнинга и экспериментов.

## Что такое AutoKeras?

AutoKeras предоставляет интерфейс высокого уровня для библиотеки Keras, делая процесс обучения моделей глубокого обучения проще и более интуитивно понятным. Он использует мощные алгоритмы оптимизации, чтобы автоматически выбрать лучшую архитектуру и параметры модели, опираясь на заданный набор данных.

## Основные модули AutoKeras:

- **Изображения**: модули для классификации изображений, распознавания объектов и других задач связанных с обработкой изображений.
- **Табличные данные**: инструменты для работы с структурированными данными, подходящие для задач регрессии и классификации.
- **Тексты**: функционал для анализа текста, включая классификацию текстов и анализ сентиментов.
- **Временные ряды**: поддержка анализа временных рядов для прогнозирования и других временно-зависимых задач.

## Введение в тюнеры и их роль в AutoKeras:

Тюнеры в AutoKeras — это компоненты, которые автоматически оптимизируют архитектуру нейронных сетей и их параметры. Они используют различные алгоритмы, такие как Bayesian Optimization, Greedy и Hyperband, чтобы найти наилучшее сочетание параметров и структуры модели. Это позволяет значительно повысить эффективность процесса обучения и достичь лучших результатов без глубоких знаний в области машинного обучения.

Используя AutoKeras, вы можете сосредоточиться на постановке задач и анализе данных, вместо того чтобы тратить время на технически сложный процесс настройки и обучения моделей.


# **3. Работа с изображениями в AutoKeras**

AutoKeras предлагает мощные инструменты для обработки и классификации изображений, делая его идеальным инструментом для решения разнообразных задач в области компьютерного зрения. В этом разделе мы подробно рассмотрим, как AutoKeras обрабатывает задачи классификации изображений.


## Как AutoKeras обрабатывает задачи классификации изображений

AutoKeras автоматизирует процесс построения и оптимизации моделей для классификации изображений. Он использует нейронные сети для анализа изображений, автоматически выбирая оптимальные архитектуры и параметры модели, что значительно упрощает задачу для разработчиков.


### Пример использования

Для классификации изображений мы можем использовать `ImageClassifier` в AutoKeras, который обрабатывает все аспекты предобработки и моделирования изображений.


In [ ]:
import autokeras as ak  # Импорт библиотеки AutoKeras

# Создание экземпляра классификатора изображений
image_classifier = ak.ImageClassifier(max_trials=3, overwrite=True)
# max_trials=3 указывает, что AutoKeras будет тестировать до трех различных моделей, чтобы найти наилучшую.
# overwrite=True говорит AutoKeras перезаписывать предыдущие неудачные попытки модели.

# Обучение модели
image_classifier.fit(x_train, y_train, epochs=10)
# x_train и y_train - это наборы данных для обучения. Наборы данных содержат изображения и их метки.
# epochs=10 указывает, что каждая модель будет обучаться на данных 10 полных проходов (эпох).

# Оценка модели
loss, accuracy = image_classifier.evaluate(x_test, y_test)
# x_test и y_test - тестовые наборы данных для оценки качества модели.
# функция возвращает значение потерь (loss) и точности (accuracy).
print("Точность классификации изображений:", accuracy)

# Экспорт и сохранение модели
model = image_classifier.export_model()  # Экспорт лучшей модели как объекта Keras.
model.save('image_classification_model', save_format='tf')  # Сохранение модели в формате TensorFlow SavedModel.
print("Модель классификации изображений сохранена в формате TensorFlow SavedModel")


В этом коде:

- **`import autokeras as ak`**: Это импортирует библиотеку AutoKeras под сокращенным названием `ak`, которое мы используем для доступа к функциональности AutoKeras.
- **`ak.ImageClassifier()`**: Создает объект классификатора изображений, который автоматически будет искать наилучшую модель для классификации изображений. Параметр `max_trials` контролирует, сколько различных моделей будет протестировано, а `overwrite` указывает на возможность перезаписи предыдущих моделей.
- **`fit()`**: Этот метод запускает процесс обучения модели на предоставленных данных. Мы указываем количество эпох обучения через параметр `epochs`.
- **`evaluate()`**: Метод оценки используется после обучения для тестирования модели на отдельном наборе данных. Он возвращает значения потерь и точности, которые дают представление о качестве модели.
- **`export_model()`**: Этот метод экспортирует обученную модель для дальнейшего использования вне AutoKeras.
- **`save()`**: Сохраняет модель на диск. Здесь мы используем формат `tf` (TensorFlow SavedModel), который является более универсальным и подходит для различных платформ и служб машинного обучения.


# **4. Работа с табличными данными в AutoKeras**

AutoKeras предоставляет удобные инструменты для работы с табличными данными, что делает его отличным выбором для задач классификации и регрессии на структурированных данных. В этом разделе мы рассмотрим, как можно использовать AutoKeras для анализа табличных данных.

## Анализ табличных данных с помощью AutoKeras

Работа с табличными данными в AutoKeras начинается с подготовки данных, что может включать предобработку, нормализацию и кодирование категориальных переменных. AutoKeras автоматизирует многие из этих процессов, позволяя пользователям сосредоточиться на интерпретации результатов, а не на монотонной предварительной обработке.


### Пример использования

Для анализа табличных данных мы можем использовать комбинацию `ak.Input()` и `RegressionHead()` или `ClassificationHead()`, в зависимости от типа задачи. Это обеспечивает гибкость в настройке модели под конкретные нужды анализа.


In [ ]:
import autokeras as ak  # Импорт библиотеки AutoKeras для автоматизированного машинного обучения

# Определение входных данных и конфигурация регрессионной задачи
input_node = ak.Input()  # Создание входного узла для модели
output_node = ak.RegressionHead(loss='mean_absolute_error')  # Добавление регрессионного слоя, используя среднюю абсолютную ошибку как функцию потерь

# Соединение входных данных с регрессионным слоем
regressor = ak.AutoModel(inputs=input_node, outputs=output_node, max_trials=10, overwrite=True)
# AutoModel автоматически тестирует различные архитектуры моделей для нахождения наилучшего решения
# max_trials=10 указывает на количество испытаний моделей
# overwrite=True позволяет перезаписывать предыдущие модели при новых испытаниях

# Обучение модели
regressor.fit(x_train, y_train, epochs=10)  # Обучение модели на тренировочных данных, x_train и y_train, на протяжении 10 эпох

# Оценка модели
loss, accuracy = model.evaluate(x_test, y_test)  # Оценка модели на тестовом наборе данных x_test и y_test
print("Точность на тестовых данных:", accuracy)  # Вывод точности модели на тестовых данных

# Экспорт и сохранение модели
exported_model = model.export_model()  # Экспортирование обученной модели
exported_model.save('table_data_model', save_format='tf')  # Сохранение модели в формате TensorFlow SavedModel
print("Модель сохранена в формате TensorFlow SavedModel как table_data_model")


В этом примере кода:
- **`import autokeras as ak`**: Импорт библиотеки AutoKeras, которая используется для автоматизированного машинного обучения, особенно полезной для пользователей без глубоких знаний в области архитектуры нейронных сетей.
- **`ak.Input()`**: Создает входной узел для модели, который служит точкой входа данных.
- **`ak.RegressionHead()`**: Добавляет регрессионный выходной слой к модели, настраивая его для использования средней абсолютной ошибки как функции потерь. Этот слой определяет, как модель будет вычислять ошибку между предсказанными и истинными значениями.
- **`ak.AutoModel()`**: Соединяет входные данные с выходным слоем и автоматически оптимизирует архитектуру модели, выполняя до 10 попыток различных конфигураций для поиска наилучшего результата.
- **`model.evaluate()`**: Этот метод оценивает, как хорошо модель работает на новых, ранее не виденных данных, предоставляя метрики, такие как потери и точность.
- **`model.export_model()`** и **`save()`**: Эти методы используются для экспорта обученной модели в стандартный формат, который может быть использован в других приложениях или сохранен для последующего использования. Модель сохраняется в формате TensorFlow SavedModel, который поддерживает все новые возможности TensorFlow и обеспечивает лучшую совместимость.

# **5. Работа с текстами в AutoKeras**

AutoKeras предлагает эффективные инструменты для работы с текстовыми данными, что делает его отличным выбором для задач классификации текста, анализа настроений и других типов текстового анализа. В этом разделе мы рассмотрим, как использовать AutoKeras для обработки текстовых данных.

## Обработка текстовых данных в AutoKeras

Обработка текстовых данных в AutoKeras начинается с предобработки текста, что включает в себя токенизацию, удаление стоп-слов, приведение слов к нормальной форме и другие техники очистки данных. AutoKeras упрощает эти задачи, автоматически оптимизируя процессы обработки в зависимости от поставленной задачи.

### Пример использования

Для задачи классификации текста мы можем использовать `TextClassifier` из AutoKeras. Этот инструмент автоматически обрабатывает текстовые данные, создает модель и оптимизирует ее параметры для лучшей производительности.

In [ ]:
import autokeras as ak  # Импорт библиотеки AutoKeras

# Создание экземпляра классификатора текста
text_classifier = ak.TextClassifier(max_trials=3, overwrite=True)
# max_trials=3 указывает, что будут проведены максимум три попытки найти наилучшую модель
# overwrite=True позволяет перезаписывать результаты предыдущих попыток

# Обучение модели
text_classifier.fit(x_train, y_train, epochs=3)
# x_train и y_train — обучающие данные и соответствующие метки
# epochs=3 означает, что модель будет обучена в течение трех эпох

# Оценка модели
loss, accuracy = text_classifier.evaluate(x_test, y_test)
# x_test и y_test — тестовые данные и метки для оценки модели
print("Точность классификации текста:", accuracy)  # Вывод точности модели на тестовых данных

# Экспорт и сохранение модели
model = text_classifier.export_model()  # Экспорт лучшей найденной модели
model.save('text_classification_model', save_format='tf')  # Сохранение модели в формате TensorFlow SavedModel
print("Модель классификации текста сохранена в формате TensorFlow SavedModel")


В этом коде:

- **`import autokeras as ak`**: Это импорт библиотеки AutoKeras, которая предоставляет инструменты для автоматической настройки и обучения моделей глубокого обучения.
- **`ak.TextClassifier()`**: Создает объект классификатора текста. Этот классификатор автоматически обрабатывает текстовые данные, выполняя такие задачи, как токенизация и векторизация текста, а также тестирование различных архитектур нейронных сетей для выбора наилучшего решения.
- **`fit()`**: Метод, который запускает процесс обучения модели на заданных данных. `epochs=3` говорит о том, что данные будут использованы для тренировки модели трижды, что помогает улучшить точность на обучающих данных.
- **`evaluate()`**: Метод, используемый для тестирования точности модели на новых, невиданных данных. Это позволяет оценить, насколько хорошо модель сможет работать в реальных условиях.
- **`export_model()`** и **`save()`**: Эти методы используются для сохранения модели после обучения, что позволяет использовать её позже или в других приложениях. Модель сохраняется в формате TensorFlow SavedModel, что обеспечивает совместимость с различными платформами и инструментами TensorFlow.

Токенизация текста в AutoKeras выполняется автоматически внутри классификатора. Токенизация — это процесс преобразования текста в последовательность токенов или слов, которые затем могут быть использованы для обучения модели. Этот процесс критически важен для обработки и понимания текста моделью.


# **6. Работа с временными рядами в AutoKeras**

AutoKeras предлагает удобные инструменты для анализа и прогнозирования временных рядов, что делает его ценным инструментом для решения задач, связанных с последовательными данными. В этом разделе мы рассмотрим, как можно использовать AutoKeras для работы с временными рядами.

## Введение в анализ временных рядов с AutoKeras

Анализ временных рядов в AutoKeras позволяет обрабатывать данные, которые представляют собой последовательности значений, изменяющихся во времени. Это может быть полезно для прогнозирования финансовых показателей, погодных условий, трендов в социальных сетях и многого другого.

### Пример использования

Для работы с временными рядами в AutoKeras мы можем использовать `TimeseriesForecaster`. Эта модель помогает автоматизировать процесс создания и тюнинга моделей для прогнозирования на основе временных данных.

In [ ]:
import autokeras as ak  # Импорт библиотеки AutoKeras для автоматизации машинного обучения

# Создание экземпляра прогнозиста временных рядов
time_series_forecaster = ak.TimeseriesForecaster(
    lookback=10,  # Количество временных шагов, которые модель рассматривает для одного прогноза
    predict_from=1,  # Начиная с какого временного шага в будущем начинать прогнозирование
    predict_until=3,  # До какого временного шага в будущем делать прогноз
    max_trials=3,  # Максимальное количество попыток для нахождения оптимальной модели
    overwrite=True  # Позволяет перезаписывать предыдущие попытки обучения
)  # Настраиваем параметры модели для обучения

# Обучение модели
time_series_forecaster.fit(x_train, y_train, epochs=5)  # Обучение модели на данных x_train с ответами y_train за 5 эпох

# Оценка модели
loss, accuracy = time_series_forecaster.evaluate(x_test, y_test)  # Оценка производительности модели на тестовых данных
print("Точность прогноза временных рядов:", accuracy)  # Вывод точности модели

# Экспорт и сохранение модели
model = time_series_forecaster.export_model()  # Экспорт обученной модели
model.save('time_series_forecaster_model', save_format='tf')  # Сохранение модели в формате TensorFlow SavedModel
print("Модель прогнозирования временных рядов сохранена в формате TensorFlow SavedModel")

В этом коде:

- **`import autokeras as ak`**: Импорт библиотеки AutoKeras, предназначенной для автоматизации процессов машинного обучения.
- **`ak.TimeseriesForecaster()`**: Эта функция создает модель для прогнозирования временных рядов. В параметрах указываются `lookback`, определяющий количество предыдущих точек данных для анализа; `predict_from` и `predict_until`, которые определяют временной интервал для прогнозирования.
- **`fit()`**: Метод `fit` используется для обучения модели на обучающих данных. `epochs=5` означает, что данные пройдут через модель пять раз, что помогает улучшить точность прогнозов.
- **`evaluate()`**: Метод `evaluate` оценивает, насколько точно модель может предсказывать данные на новом, тестовом наборе данных. Он возвращает значения потерь и точности.
- **`export_model()`** и **`save()`**: После обучения модель можно экспортировать и сохранить. Модель сохраняется в формате TensorFlow SavedModel, который поддерживает все новые возможности TensorFlow и обеспечивает лучшую совместимость для последующего использования в различных приложениях.

В предоставленном коде и описании использовалась общая форма создания модели прогнозирования временных рядов с использованием AutoKeras, которая не была специфично настроена на прогнозирование цены или тренда.


# **7. Практическая часть: Прогнозирование объема продаж**

В этом практическом разделе мы будем работать с датасетом, содержащим данные о расходах на рекламу по различным каналам (телевидение, радио, газеты) и соответствующие продажи. Наша задача — построить модель, которая сможет прогнозировать объемы продаж на основе этих расходов.


## Подготовка данных

Для начала загрузим данные и подготовим их для обработки.

In [ ]:
import pandas as pd  # Импорт библиотеки pandas для работы с данными в табличном виде
from sklearn.model_selection import train_test_split  # Импорт функции для разделения данных на обучающую и тестовую выборки

# Загрузка данных
data = pd.read_csv('/content/Data.csv')  # Чтение данных из CSV файла и сохранение в DataFrame

# Разделение данных на признаки и целевую переменную
X = data[['TV', 'radio', 'newspaper']].to_numpy()  # Извлечение признаков (столбцов) и преобразование в NumPy массив
y = data['sales'].to_numpy()  # Извлечение целевой переменной и преобразование в NumPy массив

# Разделение данных на обучающую и тестовую выборки
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
# test_size=0.2 означает, что 20% данных будут использованы как тестовая выборка
# random_state=42 обеспечивает воспроизводимость разбиения

## Настройка и обучение модели

Следующий шаг — настройка и обучение модели с использованием AutoKeras.

In [ ]:
!pip install autokeras  # Установка библиотеки AutoKeras с помощью pip в текущее среду выполнения Python

import autokeras as ak  # Импорт библиотеки AutoKeras для использования её классов и функций под сокращённым именем 'ak'

# Создание модели с помощью AutoKeras
input_node = ak.Input()  # Создание входного узла для модели, который будет принимать данные для обучения

output_node = ak.RegressionHead(metrics=['mean_absolute_error'])(input_node)
# Создание регрессионного выходного узла, который подключается к входному узлу.
# Добавление регрессионного слоя с метрикой 'mean_absolute_error'.
# 'mean_absolute_error' — это среднее значение абсолютных разностей между предсказанными и истинными значениями.
# Эта метрика показывает среднюю величину ошибки прогноза модели в процессе обучения и валидации.

auto_model = ak.AutoModel(
    inputs=input_node,  # Указание, что входные данные модели будут подаваться через созданный входной узел
    outputs=output_node,  # Указание, что вывод модели осуществляется через регрессионный выходной узел
    overwrite=True,  # Параметр, разрешающий перезапись существующих результатов модели в проекте
    max_trials=10,  # Максимальное количество попыток (триалов) для поиска оптимальной структуры модели
    tuner='random',  # Выбор стратегии поиска оптимальных параметров модели (здесь: случайный выбор)
    project_name='sales_prediction'  # Название проекта, используемое для сохранения логов и моделей в рамках данной задачи
)  # Инициализация и настройка модели AutoKeras для автоматического поиска наилучшей модели регрессии

# Обучение модели
auto_model.fit(X_train, y_train, epochs=30)
# Запуск процесса обучения модели на обучающем наборе данных X_train с соответствующими метками y_train
# Процесс обучения будет повторяться 30 раз (30 эпох), что помогает модели лучше адаптироваться и минимизировать ошибку

Trial 10 Complete [00h 00m 02s]
val_loss: 11304557568.0

Best val_loss So Far: 4.778907299041748
Total elapsed time: 00h 00m 31s
Epoch 1/30
5/5 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - loss: 872.2921 - mean_absolute_error: 22.5737   
Epoch 2/30
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 160.9778 - mean_absolute_error: 10.4037
Epoch 3/30
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 82.7331 - mean_absolute_error: 7.4378 
Epoch 4/30
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 68.3454 - mean_absolute_error: 6.5584 
Epoch 5/30
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 26.1267 - mean_absolute_error: 3.8670 
Epoch 6/30
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 7.1414 - mean_absolute_error: 2.0793 
Epoch 7/30
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 11.8627 - mean_absolute_error: 2.9166 
Epoch 8/30
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 12.5577 - mean_absolute_error: 2.8280 
Epoch 9/30
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 7.8820 - mean_absolute_error: 2.3613 
Epoch 10/30
5

## Анализ результатов

После обучения модели проведем оценку ее эффективности на тестовой выборке.

In [ ]:
# Оценка модели
_, mae = auto_model.evaluate(X_test, y_test)  # Оценка модели на тестовой выборке и получение средней абсолютной ошибки
print(f"Оценка на тестовых данных: MAE = {mae}")  # Вывод значения средней абсолютной ошибки на тестовых данных

/usr/local/lib/python3.10/dist-packages/keras/src/saving/saving_lib.py:415: UserWarning: Skipping variable loading for optimizer 'adam', because it has 2 variables whereas the saved optimizer has 6 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step - loss: 6.1182 - mean_absolute_error: 2.0253 
Оценка на тестовых данных: MAE = 2.056058406829834


После проведения серии экспериментов с моделью прогнозирования объема продаж, мы получили следующие результаты:

**Обучение модели**

Модель была обучена на 30 эпохах, в течение которых средняя абсолютная ошибка (MAE) значительно уменьшилась. Начальное значение MAE составляло 22.5737, что указывает на значительные первоначальные ошибки в прогнозах модели. Однако уже к 30-й эпохе MAE снизилась до 1.4087, что свидетельствует о значительном улучшении точности модели в процессе обучения.

**Тестирование модели**

На тестовых данных модель показала MAE равную 2.0253, что подтверждает эффективность обученной модели и её способность к генерализации результатов на новых данных. Это значение MAE является относительно низким, что делает модель достаточно надежной для использования в реальных условиях.

**Оценка модели**

Лучшая модель, полученная в ходе экспериментов, демонстрирует значение потерь val_loss, равное 4.7789, что является значительным улучшением по сравнению с первоначальными попытками (где val_loss достигала 11304557568.0). Это подчеркивает, что процесс оптимизации модели был успешным и позволил достичь высокой точности предсказаний.

**Вывод:**

Из результатов видно, что модель успешно справилась с задачей прогнозирования объемов продаж, основываясь на данных о рекламных расходах через различные каналы (TV, radio, newspaper). Уменьшение MAE в процессе обучения и низкое значение MAE на тестовой выборке свидетельствуют о высоком качестве модели. Основываясь на этих данных, можно рекомендовать модель к реальному применению для прогнозирования продаж на основе рекламных бюджетов.

# **8. Практическая часть: Классификация отзывов Tesla**

В этом разделе мы рассмотрим задачу классификации текстовых отзывов на продукцию Tesla на позитивные и негативные. Используя предоставленные данные, мы обучим модель для определения тональности отзыва.


## Обработка и подготовка текстовых данных

Первым шагом будет загрузка и подготовка данных для обучения модели.

In [ ]:
!pip install autokeras  # Установка библиотеки AutoKeras для автоматизации машинного обучения
import random  # Импорт модуля random для выполнения случайных операций, например, перемешивания данных
import tensorflow as tf  # Импорт TensorFlow для работы с тензорами и создания датасетов
from sklearn.model_selection import train_test_split  # Импорт функции для разделения данных на обучающие и тестовые наборы
from autokeras import TextClassifier  # Импорт класса TextClassifier из AutoKeras для классификации текста

# Функция для загрузки отзывов из файла
def load_reviews(file_path):
    with open(file_path, encoding='utf-8') as file:  # Открытие файла с отзывами в режиме чтения
        reviews = file.read().splitlines()  # Чтение всех строк файла и их разделение по переносам строки
    return reviews

# Загрузка данных
neg_reviews = load_reviews('/content/Негативный отзыв.txt')  # Загрузка негативных отзывов
pos_reviews = load_reviews('/content/Позитивный отзыв.txt')  # Загрузка позитивных отзывов

# Создание меток для данных
neg_labels = [0] * len(neg_reviews)  # Создание списка меток '0' для негативных отзывов
pos_labels = [1] * len(pos_reviews)  # Создание списка меток '1' для позитивных отзывов

# Объединение отзывов и меток в один список
reviews = neg_reviews + pos_reviews  # Объединение списков отзывов
labels = neg_labels + pos_labels  # Объединение списков меток

# Перемешивание данных
combined = list(zip(reviews, labels))  # Создание списка кортежей, где каждый кортеж содержит отзыв и соответствующую метку
random.shuffle(combined)  # Перемешивание списка кортежей для случайного распределения
reviews, labels = zip(*combined)  # Разделение списка кортежей обратно на отдельные списки отзывов и меток

# Разделение на обучающие и тестовые данные
X_train, X_test, y_train, y_test = train_test_split(reviews, labels, test_size=0.3)  # Разделение данных на обучающие и тестовые наборы, 30% данных для теста

# Преобразование списка в TensorFlow Dataset
train_data = tf.data.Dataset.from_tensor_slices((X_train, y_train)).batch(8)
# Метод `from_tensor_slices` создает объект `Dataset`, представляющий пары признаков (X_train) и меток (y_train), как отдельные элементы.
# Каждая пара данных будет передаваться как отдельный элемент для последующей обработки.
# Метод `batch` группирует элементы датасета в пакеты (батчи) по 8 элементов для обучения, чтобы оптимизировать производительность
# и эффективно использовать графические процессоры (GPU) или центральные процессоры (CPU) при обучении модели.

test_data = tf.data.Dataset.from_tensor_slices((X_test, y_test)).batch(8)
# Аналогичный процесс создания датасета для тестирования, который будет использоваться для оценки производительности модели.
# Группировка данных в батчи помогает быстро обрабатывать и оценивать производительность модели, загружая пакеты данных по 8 элементов.

## Настройка и обучение текстовой модели

Теперь, когда данные готовы, мы можем создать и обучить текстовую модель.

In [ ]:
# Создание текстового классификатора с AutoKeras
text_classifier = TextClassifier(max_trials=1, overwrite=True)  # Создание экземпляра классификатора текста с одной попыткой и разрешением на перезапись

# Обучение модели
text_classifier.fit(train_data, epochs=1, validation_data=test_data)  # Обучение модели на подготовленных данных, одна эпоха, с валидацией на тестовых данных

Trial 1 Complete [01h 17m 44s]
val_loss: 0.6684462428092957

Best val_loss So Far: 0.6684462428092957
Total elapsed time: 01h 17m 44s


53/53 ━━━━━━━━━━━━━━━━━━━━ 4651s 87s/step - accuracy: 0.5572 - loss: 0.7068 - val_accuracy: 0.5889 - val_loss: 0.6656


In [ ]:
# Оценка модели
print("Оценка на тестовых данных: ", text_classifier.evaluate(test_data))  # Вывод результатов оценки модели на тестовых данных

/usr/local/lib/python3.10/dist-packages/keras/src/saving/saving_lib.py:415: UserWarning: Skipping variable loading for optimizer 'adamw', because it has 1 variables whereas the saved optimizer has 403 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


23/23 ━━━━━━━━━━━━━━━━━━━━ 534s 23s/step - accuracy: 0.5931 - loss: 0.6388
Оценка на тестовых данных:  [0.6684462428092957, 0.550000011920929]


# Анализ результатов классификации отзывов Tesla

В ходе обучения модели для классификации отзывов Tesla на позитивные и негативные были достигнуты следующие результаты:

- **Общее время обучения**: Обучение модели заняло 1 час 17 минут и 44 секунды. Это значительное время, что может быть связано с использованием сложной архитектуры модели, такой как BERT.
- **Потери на обучающей выборке (loss)**: Значение потерь составило 0.7068. Это указывает на некоторые трудности модели в обучении, возможно, из-за неоптимальных гиперпараметров или несбалансированности данных.
- **Точность на обучающей выборке (accuracy)**: Точность достигла 55.72%, что является относительно низким показателем, подтверждающим, что модель могла испытывать затруднения в различении классов отзывов.
- **Потери на валидационной выборке**: Минимальные потери на валидационной выборке достигли 0.6656, что незначительно меньше, чем потери на обучающей выборке.
- **Точность на валидационной выборке**: Точность на валидационной выборке составила 58.89%, что немного лучше, чем на обучающей выборке.
- **Финальная оценка на тестовых данных**: Потери составили 0.6684, а точность — 55.00%. Эти результаты указывают на то, что модель демонстрирует схожую производительность на тренировочных, валидационных и тестовых данных.

## Выводы

Результаты показывают, что текущая модель имеет потенциал для дальнейшего улучшения, особенно в части точности классификации. Возможные шаги для улучшения включают более детальную настройку гиперпараметров, использование техник аугментации данных для увеличения обучающего набора и рассмотрение других архитектур моделей.

Также стоит обратить внимание на предупреждение о несоответствии переменных оптимизатора, что может указывать на проблемы при загрузке предобученной модели. Рекомендуется проверить совместимость версий библиотек и корректность сохранения и загрузки модели.

В целом, классификация отзывов Tesla с использованием глубокого обучения демонстрирует обнадеживающие, но не идеальные результаты, и требует дополнительных исследований и экспериментов для достижения более высокой точности и стабильности.


# **9. Практическая часть: Классификация рентгеновских снимков**

В этом разделе мы рассмотрим задачу классификации рентгеновских снимков на наличие пневмонии. Используя предоставленные медицинские изображения, мы обучим модель для автоматического определения патологий.


## Введение в медицинскую обработку изображений

Классификация медицинских изображений — это важная область, где машинное обучение может оказать значительное влияние, улучшая скорость и точность диагностики. В этом проекте мы фокусируемся на использовании сверточных нейронных сетей для анализа рентгеновских снимков грудной клетки. Сверточные нейронные сети (CNN) особенно хорошо подходят для обработки визуальных данных, таких как изображения и видео. CNN эффективно извлекают иерархические признаки из изображений, что делает их идеальным выбором для задач, связанных с изображениями, таких как распознавание и классификация объектов на изображениях, благодаря своей способности обнаруживать важные характеристики на различных уровнях


## Подготовка и анализ медицинских изображений

Прежде всего, необходимо подготовить данные для обучения модели. Изображения будут загружены, преобразованы в подходящий формат и разделены на обучающие и валидационные наборы.


In [ ]:
!pip install autokeras  # Устанавливает пакет AutoKeras с помощью инструмента pip, который является стандартным способом установки пакетов Python

import autokeras as ak  # Импортирует библиотеку AutoKeras для последующего использования в коде. 'as ak' позволяет обращаться к библиотеке с использованием сокращения 'ak'

import zipfile  # Импортирует модуль 'zipfile', который предоставляет инструменты для работы с ZIP архивами, включая их чтение и распаковку

import os  # Импортирует модуль 'os', который предоставляет множество функций для работы с операционной системой, например, для работы с файловой системой

# Открывает ZIP-файл, указанный по пути '/content/chest_xray.zip', в режиме чтения ('r')
with zipfile.ZipFile('/content/chest_xray.zip', 'r') as zip_ref:
    zip_ref.extractall('/content')  # Извлекает все файлы из указанного ZIP-архива и сохраняет их в директории '/content'

base_dir = '/content/'  # Присваивает переменной 'base_dir' строку, указывающую на базовую директорию, где находятся распакованные данные

# Создает тренировочный набор данных изображений, загружая их из директории, указанной в 'base_dir'
train_data = ak.image_dataset_from_directory(
    base_dir,  # Путь к директории с данными
    validation_split=0.2,  # Доля данных, которая будет выделена для валидационного набора
    subset="training",  # Указывает, что этот набор данных будет использоваться для тренировки
    seed=123,  # Значение для инициализации генератора случайных чисел, обеспечивает воспроизводимость разбиения данных
    image_size=(128, 128),  # Размер, к которому будут приведены все изображения
    batch_size=128,  # Количество изображений в одном пакете (батче), который подается на вход сети
    color_mode="grayscale")  # Указывает, что изображения будут загружены в градациях серого, что типично для медицинских изображений

# Создает валидационный набор данных изображений аналогичным образом
val_data = ak.image_dataset_from_directory(
    base_dir,  # Путь к директории с данными
    validation_split=0.2,  # Доля данных, которая будет выделена для валидационного набора
    subset="validation",  # Указывает, что этот набор данных будет использоваться для валидации модели
    seed=123,  # Значение для инициализации генератора случайных чисел, обеспечивает воспроизводимость разбиения данных
    image_size=(128, 128),  # Размер, к которому будут приведены все изображения
    batch_size=128,  # Количество изображений в одном пакете (батче), который подается на вход сети
    color_mode="grayscale")  # Указывает, что изображения будут загружены в градациях серого

Found 7506 files belonging to 7 classes.
Using 6005 files for training.
Found 7506 files belonging to 7 classes.
Using 1501 files for validation.


## Обучение модели и анализ результатов

Теперь мы можем обучить модель и оценить ее эффективность.

In [ ]:
from autokeras import ImageClassifier  # Импорт класса ImageClassifier из библиотеки AutoKeras

# Создание модели классификатора изображений с AutoKeras
clf = ImageClassifier(max_trials=1, overwrite=True)  # Инициализация классификатора с ограничением на количество попыток

# Запуск поиска
clf.fit(train_data, epochs=1, validation_data=val_data)  # Обучение модели на тренировочных данных и валидация на валидационных данных

Trial 1 Complete [00h 16m 40s]
val_loss: 0.07154954224824905

Best val_loss So Far: 0.07154954224824905
Total elapsed time: 00h 16m 40s
47/47 ━━━━━━━━━━━━━━━━━━━━ 753s 16s/step - accuracy: 0.7994 - loss: 0.7932 - val_accuracy: 0.9793 - val_loss: 0.0641


In [ ]:
# Оценка модели
results = clf.evaluate(val_data)
print("Оценка на тестовых данных: ", clf.evaluate(test_data))  # Вывод результатов оценки модели на тестовых данных

/usr/local/lib/python3.10/dist-packages/keras/src/saving/saving_lib.py:415: UserWarning: Skipping variable loading for optimizer 'adam', because it has 2 variables whereas the saved optimizer has 14 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


12/12 ━━━━━━━━━━━━━━━━━━━━ 79s 6s/step - accuracy: 0.9726 - loss: 0.0750
Оценка на валидационных данных:  [0.07154954224824905, 0.9753497838973999]


# Анализ результатов классификации рентгеновских снимков

В ходе обучения модели для классификации рентгеновских снимков на наличие пневмонии были достигнуты следующие результаты:

- **Время обучения**: Модель обучалась в течение 16 минут и 40 секунд, что свидетельствует о значительных вычислительных затратах, но в пределах разумного для такого рода задач.
- **Точность на обучающем наборе**: Точность составила 79.94%, что указывает на то, что модель достаточно хорошо справляется с распознаванием патологий на обучающем наборе данных.
- **Потери на обучающем наборе**: Потери составили 0.7932, что показывает, насколько хорошо модель справляется с задачей в терминах расчета ошибки между прогнозируемыми и фактическими значениями.
- **Точность на валидационном наборе**: Валидационная точность составила 97.93%, что выдающийся результат и показывает, что модель очень хорошо обобщает, распознавая патологии на новых данных.
- **Потери на валидационном наборе**: Валидационные потери составили 0.0641, подтверждая высокую эффективность модели.
- **Финальная оценка точности**: При финальной проверке на валидационном наборе точность составила 97.53%, а потери — 0.0715, что подтверждает стабильность и надежность модели.

## Выводы

Модель продемонстрировала высокую точность и низкие потери как на тренировочных, так и на валидационных данных, что делает ее надежным инструментом для помощи в медицинской диагностике пневмонии посредством рентгеновских снимков. Эти результаты подчеркивают потенциал автоматизированной диагностики в медицинской практике, способного обеспечивать быструю и точную оценку состояния пациента.

Результаты также показывают, что модель хорошо справляется с задачей минимизации ошибок и обеспечения высокой точности, что критически важно в клинических приложениях.


# **10. Итоги занятия**

Мы подошли к концу нашего занятия по AutoKeras. Надеемся, что это занятие было информативным и познавательным, и что вы смогли узнать много нового о возможностях AutoKeras и его применении для решения различных задач машинного обучения.

## Обсуждение основных уроков и выводов

В ходе занятия мы рассмотрели:
- **Основы AutoKeras**, включая его модули и функциональность для работы с изображениями, табличными данными, текстами и временными рядами.
- **Практические задачи** по классификации изображений, анализу текстов и прогнозированию временных рядов, которые показали, как можно эффективно использовать AutoKeras в реальных сценариях.
- **Оптимизацию процессов** обучения и предобработки данных для улучшения производительности и скорости моделей.

Эти знания помогут вам в дальнейшем более уверенно работать с проектами машинного обучения, а также исследовать новые возможности автоматизации и улучшения моделей.

## Вопросы и ответы


Если у вас есть какие-либо вопросы по пройденному материалу или вы хотите обсудить конкретные моменты, связанные с использованием AutoKeras в ваших проектах, пожалуйста, задайте их сейчас. Это отличная возможность углубить ваше понимание и разрешить любые сомнения.

---

Спасибо за вашу активность! Не забывайте практиковаться и экспериментировать с тем, что вы узнали сегодня. Успехов в изучении и применении машинного обучения!


# **11. Заключение и дополнительные ресурсы**

Спасибо за участие в этом занятии! Надеемся, что оно было информативным и познавательным, и вы смогли узнать больше о возможностях использования AutoKeras для различных задач машинного обучения.

## Дополнительные ресурсы

Чтобы дополнительно поддержать ваше обучение и предоставить инструменты для более глубокого изучения тем, мы предлагаем следующие ресурсы:


1. **Официальная документация AutoKeras**: Чтобы узнать больше о возможностях и API AutoKeras, посетите [официальную документацию AutoKeras](https://autokeras.com/).

2. **TensorFlow Tutorials**: Практические руководства от официального сайта TensorFlow, которые охватывают различные аспекты работы с TensorFlow, включая работу с текстами и изображениями. [TensorFlow Tutorials](https://www.tensorflow.org/tutorials).

3. **Stack Overflow**: Отличное место для поиска ответов на технические вопросы или для задания вопросов, если вы столкнулись с проблемами. [Stack Overflow](https://stackoverflow.com/).

4. **GitHub Issues для AutoKeras**: Если у вас возникли проблемы, специфичные для AutoKeras, рассмотрите возможность поиска решений или задания вопроса через систему issue на GitHub. [AutoKeras GitHub](https://github.com/keras-team/autokeras/issues).

Мы надеемся, что эти ресурсы окажутся полезными в вашем обучении и профессиональном росте. Продолжайте изучать и экспериментировать с машинным обучением, и у вас обязательно получится достичь значительных успехов в этой захватывающей области!
